concordium-consensus/src/Concordium/Scheduler.hs

```diff
@@ -83,48 +88,79 @@ import Prelude hiding (exp, mod)
   let cost = Cost.baseCost (getTransactionHeaderPayloadSize $ transactionHeader meta) (getTransactionNumSigs (transactionSignature meta))
-  unless (transactionGasAmount meta >= cost) $ throwError (Just DepositInsufficient)
   remainingBlockEnergy <- lift getRemainingEnergy
+  -- check that enough energy is remaining for the block.
   unless (remainingBlockEnergy >= cost) $ throwError Nothing
-
-  -- Now check whether the specified sender exists, and only then do all remaining checks.
-  macc <- lift (getAccount (transactionSender meta))
-  case macc of
-    Nothing -> throwError . Just $ (UnknownAccount (transactionSender meta))
-    Just iacc@(_, acc) -> do
-      amnt <- getAccountAvailableAmount acc
-      nextNonce <- getAccountNonce acc
-      let txnonce = transactionNonce meta
-      let expiry = thExpiry $ transactionHeader meta
-
-      cm <- lift getChainMetadata
-      when (transactionExpired expiry $ slotTime cm) $ throwError . Just $ ExpiredTransaction
-
-      -- After the successful credential check we check that the sender account
-      -- has enough GTU to cover the deposited energy.
+  
+  -- check that the transaction is not expired
+  cm <- lift getChainMetadata
+  when (transactionExpired (thExpiry $ transactionHeader meta) $ slotTime cm) $ throwError . Just $ ExpiredTransaction
+  
+  let addr = transactionSender meta
+  miacc <- lift (getAccount addr)
+  case miacc of
+    -- check if the account is present on the chain.
+    Nothing -> throwError (Just $ UnknownAccount addr)
+    Just iacc -> do
+      let hash = transactionHash meta
+      cache <- lift getTransactionVerificationCache
+
+      -- check that the sender account has enough funds to cover the transfer
+      amnt <- getAccountAvailableAmount $ snd iacc
       depositedAmount <- lift (energyToGtu (transactionGasAmount meta))
       unless (depositedAmount <= amnt) (throwError . Just $ InsufficientFunds)
-
-      unless (txnonce == nextNonce) (throwError . Just $ (NonSequentialNonce nextNonce))
-
-      -- Finally do the signature verification, the computationally most expensive part.
-      keys <- getAccountVerificationKeys acc
-      let sigCheck = verifyTransaction keys meta
-      unless sigCheck (throwError . Just $ IncorrectSignature)
-
-      return (iacc, cost)
+      
+      -- check if the transaction has been verified before and that the result is in the
+      -- transaction verification cache.
+      case HM.lookup hash (cache ^. normalTransactions) of
+        Just (NormalVerificationResult keysHash _) -> do
+          -- check that the keys for the account have not changed
+          currentKeys <- lift (TV.getAccountVerificationKeys $ snd iacc)
+          -- if the current account keys matches t
+          if getHash currentKeys == keysHash then return (iacc, cost)
+          -- keys changed the signature is no longer valid
+          else throwError $ Just IncorrectSignature
```

Comment:
That's not correct. It could be that the signature validates with two different sets of keys, e.g., they might have added some keys to the account but did not change the threshold.